vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,417 @@
+	subConfig, err := parseConfig(env.Config)
+	if err != nil {
+		return nil, fmt.Errorf("config parse error: %w", err)
+	}
+
+	p := &pool{
+		env:      env,
+		close:    make(chan bool, 1),
+		onClosed: make(chan error, 1),
+	}
+
+	err = p.init(params, subConfig)
+	if err != nil {
+		return nil, fmt.Errorf("can't initialize pool: %w", err)
+	}
+
+	go func() {
+		for {
+			select {
+			case <-p.close:
+				p.mu.Lock()
+				if p.proxy != nil {
+					p.proxy.terminate()
+					<-p.proxy.onTerminated
+				}
+				p.proxy = nil
+				p.onClosed <- nil
+				p.mu.Unlock()
+				return
+			case <-p.proxy.onTerminated:
+			}
+			p.mu.Lock()
+			time.Sleep(params.InitRetryDelay)
+			p.init(params, subConfig)
+			p.mu.Unlock()
+		}
+	}()
+
+	return p, nil
+}
+
+type pool struct {
+	mu       sync.Mutex
+	env      *vmimpl.Env
+	proxy    *ProxyApp
+	count    int
+	close    chan bool
+	onClosed chan error
+}
+
+func (p *pool) init(params *proxyAppParams, cfg *Config) error {
+	var err error
+	p.proxy, err = runProxyApp(params, cfg.Command)
+	if err != nil {
+		return fmt.Errorf("failed to run ProxyApp: %w", err)
+	}
+
+	count, err := p.proxy.CreatePool(string(cfg.ProxyAppConfig), p.env.Debug)
+	if err != nil || count == 0 || (p.count != 0 && p.count != count) {
+		if err == nil {
+			err = fmt.Errorf("wrong pool size %v, prev was %v", count, p.count)
+		}
+
+		p.proxy.terminate()
+		<-p.proxy.onTerminated
+		p.proxy = nil
+
+		return fmt.Errorf("failed to construct pool: %w", err)
+	}
+
+	if p.count == 0 {
+		p.count = count
+	}
+	return nil
+}
+
+func (p *pool) Count() int {
+	return p.count
+}
+
+func (p *pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	p.mu.Lock()
+	proxy := p.proxy
+	p.mu.Unlock()
+
+	if proxy == nil {
+		return nil, fmt.Errorf("can't create instance using nil pool")
+	}
+
+	return proxy.CreateInstance(workdir, index)
+}
+
+// Close is not used now. Its support require wide code changes.
+// TODO: support the pool cleanup on syz-manager level.
+func (p *pool) Close() error {
+	close(p.close)
+	return <-p.onClosed
+}
+
+// proxyAppInterface is used for tests only. But it is here to let go:generate work.
+// nolint:deadcode, unused
+type proxyAppInterface interface {
+	CreatePool(in proxyrpc.CreatePoolParams, out *proxyrpc.CreatePoolResult) error
+	CreateInstance(in proxyrpc.CreateInstanceParams, out *proxyrpc.CreateInstanceResult) error
+	Diagnose(in proxyrpc.DiagnoseParams, out *proxyrpc.DiagnoseReply) error
+	Copy(in proxyrpc.CopyParams, out *proxyrpc.CopyResult) error
+	Forward(in proxyrpc.ForwardParams, out *proxyrpc.ForwardResult) error
+	RunStart(in proxyrpc.RunStartParams, out *proxyrpc.RunStartReply) error
+	RunStop(in proxyrpc.RunStopParams, out *proxyrpc.RunStopReply) error
+	RunReadProgress(in proxyrpc.RunReadProgressParams, out *proxyrpc.RunReadProgressReply) error
+	Close(in proxyrpc.CloseParams, out *proxyrpc.CloseReply) error
+}
+
+type ProxyApp struct {
+	*rpc.Client
+	instancesCreated int32
```

Comment:
This is done in the vm package now. Remove.

Reply:
removed